app/src/main/java/org/dhis2/data/forms/dataentry/ValueStoreImpl.kt

```diff
@@ -80,14 +81,26 @@ class ValueStoreImpl(
             dataValue.attributeOptionCombo
         )
 
+        val validator = d2.dataElementModule().dataElements()
+            .uid(dataValue.dataElement).blockingGet().valueType()?.validator
+
         return if (!dataValue.value.isNullOrEmpty()) {
             if (dataValueObject.blockingExists() &&
                 dataValueObject.blockingGet().value() == dataValue.value
             ) {
                 Flowable.just(StoreResult("", ValueStoreResult.VALUE_HAS_NOT_CHANGED))
             } else {
-                dataValueObject.set(dataValue.value)
-                    .andThen(Flowable.just(StoreResult("", ValueStoreResult.VALUE_CHANGED)))
+                when (validator?.validate(dataValue.value)) {
```

Comment:
does the user has feedback about the type of error that he or she has?

Reply:
I'm not sure, but I think we only get that there is a value type error. Maybe @Balcan that worked on it, know anything else?

Reply:
I did not implement it. But it would be a good idea